geoserve/geoserve.go

```diff
@@ -101,64 +99,57 @@ func (server *GeoServer) run() {
 	for {
 		select {
 		case g := <-server.cacheGet:
-			var jsonData []byte
 			_jsonData, found := server.cache.Get(g.ip)
 			if found {
 				log.Trace("Cache hit")
-				jsonData = _jsonData.([]byte)
-			} else {
-				log.Trace("Cache miss, looking up geolocation info")
-				geoData, err := server.db.City(net.ParseIP(g.ip))
-				if err != nil {
-					log.Errorf("Unable to look up ip address %s: %s", g.ip, err)
-				} else {
-					jsonData, err = json.Marshal(geoData)
-					if err != nil {
-						log.Errorf("Unable to encode json response for ip address: %s", g.ip)
-					} else {
-						// Cache it
-						server.cache.Add(g.ip, jsonData)
-					}
-				}
+				g.resp <- _jsonData.([]byte)
+			}
+			jsonData, err := server.lookupDB(g.ip)
+			if err != nil {
+				log.Error(err)
 			}
+			server.cache.Add(g.ip, jsonData)
 			g.resp <- jsonData
-		case update := <-server.dbUpdate:
+		case db := <-server.dbUpdate:
 			if server.db != nil {
 				log.Debug("Closing old database")
 				server.db.Close()
 			}
 			log.Debug("Applying new database")
-			server.db = update.db
-			server.dbDate = update.dbDate
+			server.db = db
 			log.Debug("Clearing cached lookups")
 			server.cache = lru.New(CacheSize)
 		}
 	}
 }
 
-// keepDbCurrent checks for new versions of the database on the web every minute
-// by issuing a HEAD request.  If a new database is found, this downloads it and
-// submits it to server.dbUpdate for the run() routine to pick up.
-func (server *GeoServer) keepDbCurrent() {
+func (server *GeoServer) lookupDB(ip string) ([]byte, error) {
+	geoData, err := server.db.City(net.ParseIP(ip))
+	if err != nil {
+		return nil, fmt.Errorf("Unable to look up ip address %s: %s", ip, err)
+	}
+	jsonData, err := json.Marshal(geoData)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to encode json response for ip address: %s", ip)
+	}
+	return jsonData, nil
+}
+
+// keepDbCurrent checks the MaxMind database URL every hour and downloads it if it's
+// newer and submits it to server.dbUpdate for the run() routine to pick up.
+func (server *GeoServer) keepDbCurrent(lastModified time.Time) {
 	for {
-		time.Sleep(1 * time.Minute)
-		headResp, err := http.Head(DB_URL)
```

Comment:
It's no longer necessary as we now use `GET` with `If-Modified-Since` header. I manually verified that it works against the MaxMind URL.